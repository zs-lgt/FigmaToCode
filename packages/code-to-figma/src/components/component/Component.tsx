import * as React from 'react';
import {
    AutoLayoutProps,
    BorderProps,
    CornerProps,
    DefaultContainerProps,
    DefaultShapeProps,
    FrameSpecificProps,
    PublishableProps,
    SelectionEventProps,
    StyleOf
} from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { StyleSheet } from '../..';
import { useSelectionChange } from '../../hooks/useSelectionChange';
import {
    BorderStyleProperties,
    transformBorderStyleProperties
} from '../../styleTransformers/transformBorderProperties';
import { transformAutoLayoutToYoga } from '../../styleTransformers/transformAutoLayoutToYoga';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { useImageHash } from '../../hooks/useImageHash';
import { useNodeIdCallback } from '../../hooks/useNodeIdCallback';

export interface ComponentProps
    extends DefaultShapeProps,
        DefaultContainerProps,
        SelectionEventProps,
        AutoLayoutProps,
        BorderProps,
        CornerProps,
        FrameSpecificProps,
        OnLayoutHandlerProps,
        PublishableProps {
    style?: StyleOf<
        GeometryStyleProperties &
            YogaStyleProperties &
            LayoutStyleProperties &
            BlendStyleProperties &
            BorderStyleProperties
    >;
    nodeRef?: any;
}

const Component: React.FC<ComponentProps> = props => {
    const nodeRef = props.nodeRef || React.useRef();
    useSelectionChange(nodeRef, props);
    useNodeIdCallback(nodeRef, props.onNodeId);
    const style = { ...StyleSheet.flatten(props.style), ...transformAutoLayoutToYoga(props) };
    const imageHash = useImageHash(style.backgroundImage);
    const componentProps = {
        ...transformLayoutStyleProperties(style),
        ...transformBlendProperties(style),
        ...transformGeometryStyleProperties('backgrounds', style, imageHash),
        ...transformBorderStyleProperties(style),
        ...props,
        style
    };
    const yogaChildProps = useYogaLayout({ nodeRef, ...componentProps });

    useOnLayoutHandler(yogaChildProps, props);

    return <component {...componentProps} {...yogaChildProps} innerRef={nodeRef} />;
};

export { Component };
