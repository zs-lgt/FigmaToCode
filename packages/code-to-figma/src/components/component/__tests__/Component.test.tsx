import * as React from 'react';
import { act, create } from 'react-test-renderer';
import { Component } from '../Component';
import { createFigma } from 'figma-api-stub';

describe('<Component />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Component without props', () => {
        let tree;
        act(() => {
            tree = create(<Component />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('Component with layout style properties', () => {
        let tree;
        act(() => {
            tree = create(
                <Component
                    style={{
                        width: 100,
                        height: 200,
                        top: 50,
                        left: 20
                    }}
                />
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('Component with color background', () => {
        let tree;
        act(() => {
            tree = create(
                <Component
                    style={{
                        backgroundColor: '#1bff00'
                    }}
                />
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('Component with image background', () => {
        let tree;
        act(() => {
            tree = create(
                <Component
                    style={{
                        backgroundColor: '#ff8b29',
                        backgroundSize: 'cover',
                        backgroundImage: './image.png'
                    }}
                />
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('Component with border', () => {
        let tree;
        act(() => {
            tree = create(
                <Component
                    style={{
                        borderRadius: 10,
                        borderWidth: 1,
                        borderColor: '#222020'
                    }}
                />
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('Component with auto layout', () => {
        let tree;
        act(() => {
            tree = create(
                <Component
                    layoutMode="HORIZONTAL"
                    horizontalPadding={4}
                    verticalPadding={1}
                    style={{
                        borderRadius: 10,
                        borderWidth: 1,
                        borderColor: '#222020'
                    }}
                />
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
