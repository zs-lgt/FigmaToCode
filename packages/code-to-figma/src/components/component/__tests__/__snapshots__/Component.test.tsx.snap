// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Component /> Component with auto layout 1`] = `
<component
  cornerRadius={10}
  horizontalPadding={4}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  layoutMode="HORIZONTAL"
  strokeWeight={1}
  strokes={
    Array [
      Object {
        "color": Object {
          "b": 0.12549019607843137,
          "g": 0.12549019607843137,
          "r": 0.13333333333333333,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  style={
    Object {
      "borderColor": "#222020",
      "borderRadius": 10,
      "borderWidth": 1,
      "flexDirection": "row",
      "paddingBottom": 1,
      "paddingLeft": 4,
      "paddingRight": 4,
      "paddingTop": 1,
    }
  }
  verticalPadding={1}
/>
`;

exports[`<Component /> Component with border 1`] = `
<component
  cornerRadius={10}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  strokeWeight={1}
  strokes={
    Array [
      Object {
        "color": Object {
          "b": 0.12549019607843137,
          "g": 0.12549019607843137,
          "r": 0.13333333333333333,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  style={
    Object {
      "borderColor": "#222020",
      "borderRadius": 10,
      "borderWidth": 1,
    }
  }
/>
`;

exports[`<Component /> Component with color background 1`] = `
<component
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 1,
          "r": 0.10588235294117647,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#1bff00",
    }
  }
/>
`;

exports[`<Component /> Component with image background 1`] = `
<component
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff8b29",
      "backgroundImage": "./image.png",
      "backgroundSize": "cover",
    }
  }
/>
`;

exports[`<Component /> Component with layout style properties 1`] = `
<component
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "height": 200,
      "left": 20,
      "top": 50,
      "width": 100,
    }
  }
  width={100}
  x={20}
  y={50}
/>
`;

exports[`<Component /> Component without props 1`] = `
<component
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
/>
`;
